### DISCUSSION TITLE
How to configure private registry for scoped packages?

### DISCUSSION THREAD

**Discussion by himanshumehta1114** [renovatebot/renovate#9965] [Request Help] [answered]
Hi Folks!

I have few packages on [self-hosted npm instance](https://verdaccio.org/). After I updated renovate to v25 on self-hosted GitLab. The documentation doesn't recommend using the `.npmrc` file to configure the private registry. How can I configure renovate to lookup for scoped packages on the private registry without using `.npmrc`?

I tried using hostRules in the packageRules, but it didn't work.

I'm using the following configuration:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    "group:allNonMajor",
    ":automergePatch",
    ":rebaseStalePrs"
  ],
  "rangeStrategy": "bump",
  "dependencyDashboard": true,
  "gitLabAutomerge": true,
  "packageRules": [
    {
      "description": "Automerge @scope minor and patch updates",
      "matchPackagePrefixes": ["@scope"],
      "groupName": "scoped deps",
      "hostRules": [{ hostName: "npm.mycompany.io" }]
    }
  ]
}
```

**Comment by rarkins** [comment, marked as answer]
Option 1: Copy the contents of your `.npmrc` into your bot's config (in the `npmrc` field) or via env - `RENOVATE_NPMRC`.

Better would be to commit a `.npmrc` to the repo containing `@scope:registry=https://npm.mycompany.io`. This will be understood by Renovate as well as clients. Then you can add a simple `hostRules` entry just for the auth.

FYI having a `hostRules` value inside `packageRules` is not valid.

**Reply by himanshumehta1114** [reply]
@rarkins this works, thanks for your reply. Same I will have to do for yarn2 as well? Right now I see it doesn't detect yarn2 as a different package manager. So registries from `.yarnrc.yml` file are not read. Any way I can configure renovate to use `yarnrc.yml` ?

**Reply by rarkins** [reply]
For yarn2, best if you:
- Commit the `.yarnrc.yml` file with registries in the repo (you probably have it already.. yarn 2 needs it)
- Also add a package rule to Renovate config so that Renovate knows where to look:

```
  "packageRules": [
    {
      "matchPackagePrefixes": ["@scope"],
      "registryUrls": ["npm.mycompany.io"]
    }
  ]
```

But we should ideally detect the registry rules in `.yarnrc.yml` automatically in future. If you can product an example repo and create a feature request for that (in an Issue) then it would help us to move it forward. It's ok if the registry you point to is made up/unreachable.

**Reply by himanshumehta1114** [reply]
Sure, I will create a sample repo with yarn2 & create a feature request for extracting registry from `yarnrc.yml`.

I tried using `registryUrls`:

```
"packageRules": [
    {
      "description": "Automerge @scope minor and patch updates",
      "matchPackagePrefixes": ["@scope"],
      "groupName": "scope deps",
      "groupSlug": "scope-deps",
      "registryUrls": ["https://npm.mycompany.io"]
    }
  ]
```

But I get the following warning
![image](https://user-images.githubusercontent.com/25463496/117938024-3a1b6300-b324-11eb-8529-c55489489148.png)

**Reply by rarkins** [reply]
Oh, sorry.. I forgot about that. We plan to improve the logic so that it works soon.

Can you try adding `"npmrc": ["@scope:registry=https://npm.mycompany.io"]` to your config instead of the `registryUrls`? It will override any `.npmrc` in the repo but for a Yarn repo that shouldn't be a problem

**Reply by himanshumehta1114** [reply]
Sure, thanks 🙂

**Comment by bytestream** [comment]
> Better would be to commit a `.npmrc` to the repo containing `@scope:registry=https://npm.mycompany.io`. This will be understood by Renovate as well as clients. Then you can add a simple `hostRules` entry just for the auth.

@rarkins This works for the root `package.json` file but does this work for `package.json` in sub directories? It worked in v24 but in v25 scoped dependencies in `package.json` sub directories produces `Dependency lookup failure: not found`

**Reply by rarkins** [reply]
.npmrc in subdirs should be detected

**Reply by bytestream** [reply]
There is only one `.npmrc` in the root:
```
/.npmrc
/package.json
/dir1/
/dir1/package.json
```

`/dir1/package.json` is where the problem is

**Reply by rarkins** [reply]
Npm doesn't "hunt" up directory structures so neither does Renovate. Are you using yarn workspaces though?

**Reply by bytestream** [reply]
It's odd because that error didn't exist in the v24 logs. I'm using `npm` not `yarn`.

**Reply by rarkins** [reply]
Npm will look in the current directory and then in the home directory. If you configured ~/.npmrc then it would explain what you saw.
